coaster-nn/src/frameworks/native/helper.rs

```diff
@@ -307,17 +307,32 @@ macro_rules! impl_ops_softmax_for {
                 x: &SharedTensor<$t>,
                 result: &mut SharedTensor<$t>,
             ) -> Result<(), Error> {
+                // Input tensor must have at least 2 dimensions.
+                // First dimension is treated as a batch number.
+                assert!(
+                    x.desc().size() > 1,
+                    "Input tensor for softmax must have at least 2 dimensions, got {:?}",
+                    x.desc()
+                );
+
+                let batch_size = x.desc()[0];
+                let item_size = x.desc().iter().skip(1).fold(1, |acc, v| acc * v);
+
                 let xs = read!(x, $t, self);
                 let rs = write_only!(result, $t, self);
 
                 map1(xs, rs, |v| v.exp())?;
 
-                let mut sum: $t = 0.0; // iter_arith is not stable yet
-                for r in &*rs {
-                    sum += *r;
-                }
-                for r in rs {
-                    *r /= sum;
+                for i in 0..batch_size {
+                    let batch_item = &mut rs[i * item_size..(i + 1) * item_size];
```

Comment:
Could be expressed as `&mut rs[i * item_size][..item_size]`

Reply:
Oh? I've never seen this syntax. And compiler thinks that's an error too:
```
error[E0608]: cannot index into a value of type `f32`
   --> coaster-nn/src/frameworks/native/helper.rs:355:42
    |
355 |                     let batch_item_in = &xs[i * item_size][..item_size];
    |                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Mean xulpmea culpa, `&slic[start..][..length]` is the right syntax, missed the dots in the first index operator

Reply:
Thanks, didn't know about this syntax. Done.